File: ugc/src/main/java/com/inappstory/sdk/ugc/picker/FilePreviewsAdapter.java
package com.inappstory.sdk.ugc.picker;

import android.annotation.SuppressLint;
import android.content.Context;
import android.util.Pair;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.inappstory.sdk.stories.utils.Sizes;
import com.inappstory.sdk.ugc.R;

import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Date;
import java.util.Iterator;
import java.util.List;

class FilePreviewsAdapter extends RecyclerView.Adapter<FilePreviewsHolder> {
    FilePicker picker;
    List<FilePicker.FileData> previews = new ArrayList<>();
    FilePreviewsCache cache = new FilePreviewsCache();
    OpenCameraClickCallback cameraCallback;
    FileClickCallback clickCallback;
    NoAccessCallback noAccessCallback;
    boolean hasFileAccess;
    boolean allowMultipleSelection;
    String galleryAccessText;
    int galleryFileMaxCount;
    String galleryFileLimitText;

    public FilePreviewsAdapter(Context context,
                               boolean hasFileAccess,
                               boolean allowMultipleSelection,
                               List<String> mimeTypes,
                               FileClickCallback clickCallback,
                               OpenCameraClickCallback cameraCallback,
                               NoAccessCallback noAccessCallback,
                               String galleryAccessText,
                               int galleryFileMaxCount,
                               String galleryFileLimitText) {
        this.noAccessCallback = noAccessCallback;
        this.galleryFileMaxCount = galleryFileMaxCount;
        this.galleryFileLimitText = galleryFileLimitText;
        this.galleryAccessText = galleryAccessText;
        this.cameraCallback = cameraCallback;
        this.clickCallback = clickCallback;
        this.hasFileAccess = hasFileAccess;
        this.allowMultipleSelection = allowMultipleSelection;
        this.picker = new PhotoVideoPicker();
        if (hasFileAccess) {
            previews.addAll(
                    picker.getImagesPath(
                            context,
                            new PickerFilter(30000000L, 30000L),
                            mimeTypes
                    )
            );
        }
    }


    @Override
    public void onViewRecycled(@NonNull FilePreviewsHolder holder) {
        super.onViewRecycled(holder);
        if (holder.path != null && cache != null) {
            cache.remove(holder.path);
        }
    }

    @NonNull
    @Override
    public FilePreviewsHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View v;
        if (viewType == -1) {
            v = LayoutInflater.from(parent.getContext()).inflate(R.layout.cs_file_camera_cell,
                    parent, false);
        } else if (viewType == -2) {
            v = LayoutInflater.from(parent.getContext()).inflate(R.layout.cs_file_no_access_cell,
                    parent, false);
        } else {
            v = LayoutInflater.from(parent.getContext()).inflate(R.layout.cs_file_picker_cell,
                    parent, false);

        }
        v.getLayoutParams().width = Sizes.getScreenSize(v.getContext()).x / 3;
        v.getLayoutParams().height = (16 * v.getLayoutParams().width) / 9;

        return new FilePreviewsHolder(v);
    }

    @Override
    public int getItemViewType(int position) {
        if (position == 0) return -1;
        if (!hasFileAccess && position == 1) return -2;
        return position;
    }

    @Override
    public long getItemId(int position) {
        if (position == 0) return -1;
        return position;
    }

    public List<Integer> activePositions = new ArrayList<>();

    @SuppressLint("DefaultLocale")
    private String convertLongToTime(long seconds) {
        long s = seconds % 60;
        long m = (seconds / 60) % 60;
        long h = (seconds / (60 * 60)) % 24;
        return seconds >= 3600 ? String.format("%02d:%02d:%02d", h, m, s) : String.format("%02d:%02d", m, s);
    }

    @Override
    public void onBindViewHolder(@NonNull FilePreviewsHolder holder, int position) {
        Integer intPos = Integer.valueOf(position - 1);
        if (position != 0) {
            if (hasFileAccess) {
                FilePicker.FileData data = previews.get(position - 1);
                String path = data.component1();
                Long duration = data.component2();
                holder.itemView.setSelected(activePositions.contains(intPos));
                ImageView iv = holder.itemView.findViewById(R.id.image);
                if (iv != null) {
                    holder.path = path;
                    cache.loadPreview(path, iv, duration != null);
                }
                TextView count = holder.itemView.findViewById(R.id.count);
                TextView videoDuration = holder.itemView.findViewById(R.id.videoDuration);
                if (duration != null) {
                    videoDuration.setVisibility(View.VISIBLE);
                    videoDuration.setText(convertLongToTime(duration / 1000));
                } else {
                    videoDuration.setVisibility(View.GONE);
                }
                if (activePositions.contains(intPos)) {
                    count.setText(Integer.toString(activePositions.indexOf(position - 1) + 1));
                    count.setVisibility(View.VISIBLE);
                } else {
                    count.setVisibility(View.GONE);
                }
                holder.itemView.setOnClickListener(v -> {
                    if (activePositions.contains(intPos)) {
                        clickCallback.unselect(path);
                        activePositions.remove(intPos);
                    } else {
                        if (activePositions.size() >= galleryFileMaxCount) {
                            Toast.makeText(
                                    holder.itemView.getContext(),
                                    galleryFileLimitText,
                                    Toast.LENGTH_SHORT
                            ).show();
                            return;
                        }
                        activePositions.add(intPos);
                        clickCallback.select(path);
                        if (!allowMultipleSelection) {
                            Iterator<Integer> i = activePositions.iterator();
                            while (i.hasNext()) {
                                Integer activePosition = i.next();
                                if (activePosition.intValue() != intPos.intValue()) {
                                    clickCallback.unselect(previews.get(activePosition).component1());
                                    i.remove();
                                }
                            }
                        }
                    }
                    notifyDataSetChanged();
                });
            } else {
                TextView text = holder.itemView.findViewById(R.id.gallery_access_text);
                text.setText(galleryAccessText);
                holder.itemView.setOnClickListener(v -> {
                    noAccessCallback.click();
                });
            }
        } else {
            holder.itemView.setOnClickListener(v -> {
                cameraCallback.open();
            });
        }
    }


    @Override
    public int getItemCount() {
        return (hasFileAccess ? previews.size() : 1) + 1;
    }
}

File: ugc/src/main/java/com/inappstory/sdk/ugc/picker/FilePreviewsAdapter.kt
package com.inappstory.sdk.ugc.picker

import android.annotation.SuppressLint
import android.content.Context
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import android.widget.Toast
import androidx.recyclerview.widget.RecyclerView
import com.inappstory.sdk.stories.utils.Sizes
import com.inappstory.sdk.ugc.R

internal class FilePreviewsAdapter : RecyclerView.Adapter<FilePreviewsHolder> {

    var picker: FilePicker
    private val previews = arrayListOf<FilePicker.FileData>()
    private val cache = FilePreviewsCache(false)
    private val cameraCallback: OpenCameraClickCallback
    private val storageCallback: OpenStorageClickCallback
    private val clickCallback: FileClickCallback
    private val noAccessCallback: NoAccessCallback
    private var hasFileAccess = false
    private var allowMultipleSelection = false
    private var galleryFileMaxCount = 0
    private val translations: MutableMap<String, String> = hashMapOf()

    constructor(
        context: Context?,
        hasFileAccess: Boolean,
        allowMultipleSelection: Boolean,
        mimeTypes: List<String>,
        clickCallback: FileClickCallback,
        cameraCallback: OpenCameraClickCallback,
        storageCallback: OpenStorageClickCallback,
        noAccessCallback: NoAccessCallback,
        galleryFileMaxCount: Int,
        translations: Map<String, String>,
        pickerFilter: PickerFilter
    ) {
        this.noAccessCallback = noAccessCallback
        this.translations.putAll(translations)
        this.galleryFileMaxCount = galleryFileMaxCount
        this.cameraCallback = cameraCallback
        this.storageCallback = storageCallback
        this.clickCallback = clickCallback
        this.hasFileAccess = hasFileAccess
        this.allowMultipleSelection = allowMultipleSelection
        picker = PhotoVideoPicker()
        if (hasFileAccess) {
            previews.addAll(
                picker.getImagesPath(
                    context!!,
                    pickerFilter,
                    mimeTypes
                )
            )
        }
    }

    override fun onViewRecycled(holder: FilePreviewsHolder) {
        super.onViewRecycled(holder)
        if (holder.path != null) {
            cache.remove(holder.path)
        }
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): FilePreviewsHolder {
        val v = when (viewType) {
            -1 -> LayoutInflater.from(parent.context).inflate(
                R.layout.cs_file_camera_cell,
                parent, false
            )

            -2 -> LayoutInflater.from(parent.context).inflate(
                R.layout.cs_file_storage_cell,
                parent, false
            )

            -3 -> LayoutInflater.from(parent.context).inflate(
                R.layout.cs_file_no_access_cell,
                parent, false
            )

            else -> LayoutInflater.from(parent.context).inflate(
                R.layout.cs_file_picker_cell,
                parent, false
            )
        }
        v.layoutParams.width = Sizes.getScreenSize(v.context).x / 3
        v.layoutParams.height = 16 * v.layoutParams.width / 9
        return FilePreviewsHolder(v)
    }

    private val activePositions: ArrayList<Int> = arrayListOf()

    override fun onBindViewHolder(holder: FilePreviewsHolder, position: Int) {
        val intPos = Integer.valueOf(position - 2)
        if (position > 1) {
            if (hasFileAccess) {
                val file = previews[position - 1]
                val path = file.name
                val duration = file.duration
                holder.itemView.isSelected = activePositions.contains(intPos)
                val iv = holder.itemView.findViewById<ImageView>(R.id.image)
                if (iv != null) {
                    holder.path = path
                    cache.loadPreview(
                        path,
                        iv,
                        file.unavailableByDuration || file.unavailableBySize,
                        duration != null
                    )
                }
                val count = holder.itemView.findViewById<TextView>(R.id.count)
                val videoDuration = holder.itemView.findViewById<TextView>(R.id.videoDuration)
                if (duration != null) {
                    videoDuration.visibility = View.VISIBLE
                    videoDuration.text = convertLongToTime(duration / 1000)
                } else {
                    videoDuration.visibility = View.GONE
                }
                if (activePositions.contains(intPos)) {
                    count.text = "${(activePositions.indexOf(position - 1) + 1)}"
                    count.visibility = View.VISIBLE
                } else {
                    count.visibility = View.GONE
                }
                if (file.unavailableByDuration || file.unavailableBySize) {
                    count.text = "!"
                    count.visibility = View.VISIBLE
                }
                holder.itemView.setOnClickListener {
                    if (file.unavailableByDuration) {
                        showToast(
                            holder.itemView.context,
                            translations["fileLimitVideoDuration"] ?: ""
                        )
                    } else if (file.unavailableBySize) {
                        showToast(
                            holder.itemView.context,
                            (if (file.type == "video") {
                                translations["fileLimitVideoSize"]
                            } else {
                                translations["fileLimitPhotoSize"]
                            }) ?: ""
                        )
                    } else if (activePositions.contains(intPos)) {
                        clickCallback.unselect(SelectedFile(path, file.type))
                        activePositions.remove(intPos)
                    } else {
                        if (activePositions.size >= galleryFileMaxCount) {
                            showToast(
                                holder.itemView.context,
                                translations["galleryFileLimitText"] ?: ""
                            )
                            return@setOnClickListener
                        }
                        activePositions.add(intPos)
                        Log.d("PhotoPicker", "Old $path")
                        clickCallback.select(SelectedFile(path, file.type))
                        if (!allowMultipleSelection) {
                            val i: MutableIterator<Int> = activePositions.iterator()
                            while (i.hasNext()) {
                                val activePosition = i.next()
                                if (activePosition != intPos.toInt()) {
                                    clickCallback.unselect(
                                        SelectedFile(
                                            previews[activePosition].name,
                                            previews[activePosition].type
                                        )
                                    )
                                    i.remove()
                                }
                            }
                        }
                    }
                    notifyChanges()
                }
            } else {
                val text = holder.itemView.findViewById<TextView>(R.id.gallery_access_text)
                text.text = translations["galleryAccessText"] ?: ""
                holder.itemView.setOnClickListener { noAccessCallback.click() }
            }
        } else if (position == 0) {
            holder.itemView.setOnClickListener { cameraCallback.open() }
        } else {
            holder.itemView.setOnClickListener { storageCallback.open() }
        }
    }

    private fun showToast(context: Context, text: String) {
        Toast.makeText(
            context,
            text,
            Toast.LENGTH_SHORT
        ).show()
    }

    @SuppressLint("DefaultLocale")
    private fun convertLongToTime(seconds: Long): String {
        val s = seconds % 60
        val m = seconds / 60 % 60
        val h = seconds / (60 * 60) % 24
        return if (seconds >= 3600)
            String.format("%02d:%02d:%02d", h, m, s)
        else
            String.format("%02d:%02d", m, s)
    }

    private fun notifyChanges() {
        notifyDataSetChanged()
    }

    override fun getItemViewType(position: Int): Int =
        if (position == 0) -1
        else if (position == 1) -2
        else if (!hasFileAccess && position == 2) -3
        else position

    override fun getItemId(position: Int): Long =
        when (position) {
            0 -> -1
            1 -> -2
            else -> position.toLong()
        }

    override fun getItemCount(): Int {
        return (if (hasFileAccess) previews.size else 1) + 2
    }
}